File: src/main/java/kitchenpos/order/tobe/eatinorder/application/OrderTableService.java
package kitchenpos.order.tobe.eatinorder.application;


import kitchenpos.order.tobe.eatinorder.application.dto.request.OrderTableChangeNumberRequest;
import kitchenpos.order.tobe.eatinorder.application.dto.request.OrderTableCreateRequest;
import kitchenpos.order.tobe.eatinorder.application.dto.response.OrderTableResponse;
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderRepository;
import kitchenpos.order.tobe.eatinorder.domain.EatInOrderStatus;
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTable;
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTableRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

@Service("newOrderTableService")
public class OrderTableService {
    private final EatInOrderRepository eatInOrderRepository;
    private final OrderTableRepository orderTableRepository;

    public OrderTableService(EatInOrderRepository eatInOrderRepository, OrderTableRepository orderTableRepository) {
        this.eatInOrderRepository = eatInOrderRepository;
        this.orderTableRepository = orderTableRepository;
    }

    @Transactional
    public OrderTableResponse create(final OrderTableCreateRequest request) {
        final OrderTable orderTable = new OrderTable(UUID.randomUUID(), request.name());

        return toOrderTableResponse(orderTableRepository.save(orderTable));
    }

    @Transactional
    public OrderTableResponse sit(final UUID orderTableId) {
        final OrderTable orderTable = findOrderTableById(orderTableId);

        orderTable.sit();

        return toOrderTableResponse(orderTable);
    }

    @Transactional
    public OrderTableResponse clear(final UUID orderTableId) {
        final OrderTable orderTable = findOrderTableById(orderTableId);

        if (eatInOrderRepository.existsByOrderTableAndStatusNot(orderTable, EatInOrderStatus.COMPLETED)) {
            throw new IllegalStateException("주문이 완료되지 않은 테이블은 초기화할 수 없습니다.");
        }

        orderTable.clear();

        return toOrderTableResponse(orderTable);
    }

    @Transactional
    public OrderTableResponse changeNumberOfGuests(final UUID orderTableId, final OrderTableChangeNumberRequest request) {
        final int numberOfGuests = request.numberOfGuest();
        final OrderTable orderTable = findOrderTableById(orderTableId);

        orderTable.changeNumberOfGuests(numberOfGuests);

        return toOrderTableResponse(orderTable);
    }

    @Transactional(readOnly = true)
    public List<OrderTableResponse> findAll() {
        return orderTableRepository.findAll().stream()
                .map(this::toOrderTableResponse)
                .toList();
    }

    private OrderTable findOrderTableById(final UUID orderTableId) {
        return orderTableRepository.findById(orderTableId)
                .orElseThrow(() -> new NoSuchElementException("주문 테이블을 찾을 수 없습니다."));
    }

    private OrderTableResponse toOrderTableResponse(OrderTable orderTable) {
        return new OrderTableResponse(
                orderTable.getId(),
                orderTable.getName(),
                orderTable.getNumberOfGuests(),
                orderTable.isOccupied()
        );
    }
}

File: src/main/kotlin/kitchenpos/order/tobe/eatinorder/application/OrderTableService.kt
package kitchenpos.order.tobe.eatinorder.application

import java.util.*
import kitchenpos.order.tobe.eatinorder.application.dto.CreateOrderTableReq
import kitchenpos.order.tobe.eatinorder.domain.OrderTable
import kitchenpos.order.tobe.eatinorder.domain.OrderTableEmptyService
import kitchenpos.order.tobe.eatinorder.domain.OrderTableName
import kitchenpos.order.tobe.eatinorder.domain.OrderTableOccupancy
import kitchenpos.order.tobe.eatinorder.domain.OrderTableRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service("tobeOrderTableService")
class OrderTableService(
    private val orderTableRepository: OrderTableRepository,
    private val orderTableEmptyService: OrderTableEmptyService
) {

    @Transactional
    fun create(req: CreateOrderTableReq): UUID {
        return orderTableRepository.save(
            OrderTable(
                orderTableName = OrderTableName(req.name),
                orderTableOccupancy = OrderTableOccupancy.EMPTY
            )
        ).id
    }

    @Transactional
    fun occupied(orderTableId: UUID) {
        val orderTable =
            orderTableRepository.findById(orderTableId).orElseThrow { NoSuchElementException("주문 테이블을 찾을 수 없습니다.") }
        orderTable.occupied()
    }

    @Transactional
    fun empty(orderTableId: UUID) {
        val orderTable =
            orderTableRepository.findById(orderTableId).orElseThrow { NoSuchElementException("주문 테이블을 찾을 수 없습니다.") }
        orderTable.empty(orderTableEmptyService)
    }

    @Transactional
    fun changeNumberOfGuest(orderTableId: UUID, numberOfGuest: Int) {
        val orderTable =
            orderTableRepository.findById(orderTableId).orElseThrow { NoSuchElementException("주문 테이블을 찾을 수 없습니다.") }
        orderTable.changeNumberOfGuest(numberOfGuest)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<OrderTable> {
        return orderTableRepository.findAll()
    }
}
